### PR TITLE
Error out when `/dev/mqueue/` doesn't exist and container engine is `podman` with EE support enabled

### PR DESCRIPTION
**Pull request by cidrblock**
In the case of WSL2 + Ubuntu, no `/dev/mqueue` directory is present, but runner provides `ipc=host` to podman which requires POSIX messaage queue support.

Related: #610 
Related: https://github.com/ansible/ansible-runner/issues/984

Tests added.

### CODE REVIEW

File: docs/changelog-fragments.d/866.bugfix.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+Detect and error for missing `/dev/mqueue` directory when
```
Comment (webknjaz) [inline docs/changelog-fragments.d/866.bugfix.md:1]:
```suggestion
Started erroring out loudly on missing `/dev/mqueue/` directory when
```
(past tense for consistency —see the RTD preview + trailing slash to emphasize that it's a dir)

File: docs/changelog-fragments.d/866.bugfix.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+execution environment support is enabled and the container
+engine is set to `podman`. This is a condition commonly
+found using Ubuntu with WSL2.
+-- by {user}`cidrblock`
```
Comment (webknjaz) [inline docs/changelog-fragments.d/866.bugfix.md:5]: IIRC Markdown would inline this into the paragraph in there's no empty line
```suggestion

-- by {user}`cidrblock`
```

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py [hunk 3]
```diff
@@ -281,6 +281,20 @@ def execution_environment(self, entry: SettingsEntry, config) -> PostProcessorRe
                         )
                     if hint:
                         exit_messages.append(ExitMessage(message=hint, prefix=ExitPrefix.HINT))
+
+            # check for /dev/mqueue when using podman because runner passes ipc=host
+            # https://github.com/ansible/ansible-navigator/issues/610
+            if config.container_engine == "podman" and not Path("/dev/mqueue").exists():
```
Comment (webknjaz) [inline src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:287]: Maybe use is_dir since the error message claims exactly that?

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py [hunk 4]
```diff
@@ -281,6 +281,20 @@ def execution_environment(self, entry: SettingsEntry, config) -> PostProcessorRe
...
+                exit_msg = (
+                    "Execution environment support while using podman requires a"
+                    " '/dev/mqueue' directory."
```
Comment (webknjaz) [inline src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:290]: Same as with the change note
```suggestion
                    " '/dev/mqueue/' directory."
```

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py [hunk 5]
```diff
@@ -281,6 +281,20 @@ def execution_environment(self, entry: SettingsEntry, config) -> PostProcessorRe
...
+                )
+                exit_messages.append(ExitMessage(message=exit_msg))
+                hint = (
+                    "Try creating it with 'mkdir /dev/mqueue' or reference the documentation for"
```
Comment (webknjaz) [inline src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:294]: I think it may be reasonable to put this path into a module-level constant
```suggestion
                    "Try creating it with 'mkdir /dev/mqueue/' or reference the documentation for"
```

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+"""Tests related to a missing ``/dev/mqueue`` directory when using ``podman``."""
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:1]:
```suggestion
"""Tests related to a missing ``/dev/mqueue/`` directory when using ``podman``."""
```

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    """Return false when checking for a file or directory path.
+
+    :param _arg: The file or directory path to look for
+    :returns: False, indicating it does not exist
+    """
+    return False
+
+
+def test_posix_message_queue_ee(monkeypatch, generate_config):
+    """Check for error messages related to missing ``/dev/mqueue`` and ``podman`` when ee true.
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:15]:
```suggestion
    """Check for error messages related to missing ``/dev/mqueue/`` and ``podman`` when ee true.
```

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    :param monkeypatch: The monkeypatch fixture
+    :param generate_config: The configuration generator fixture
+    """
+    monkeypatch.setattr(Path, "exists", exists)
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:20]: Since this is copy-pasted in all the tests, I recommend putting it into a local fixture.

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
+"""Tests related to a missing ``/dev/mqueue`` directory when using ``podman``."""
+from pathlib import Path
+
+
+def exists(_arg):
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:5]: maybe rename this to

```suggestion
def _never_exists(_arg):
```

**Review by webknjaz — COMMENTED** [review]
Maybe also add a note to the FAQ document?

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 7 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: tests were completely rewritten....

Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: got it! ty

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 8 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: tests were completely rewritten....

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 9 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: tests were completely rewritten....

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 6 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: got it TY!

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py (hunk 5 above)
Comment (cidrblock) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py]: used a var so it's the same everywhere, good idea

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py (hunk 4 above)
Comment (cidrblock) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py]: used a var so it's consistent, thx

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py (hunk 3 above)
Comment (cidrblock) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py]: added it, good idea

File: docs/changelog-fragments.d/866.bugfix.md (hunk 2 above)
Comment (cidrblock) [reply docs/changelog-fragments.d/866.bugfix.md:7]: added, looks better

File: docs/changelog-fragments.d/866.bugfix.md (hunk 1 above)
Comment (cidrblock) [reply docs/changelog-fragments.d/866.bugfix.md]: rephrased. lmk

**Review by ganeshrn — APPROVED** [review]
I think the work that @webknjaz is doing with enabling windows + WSL2 in CI for `ansible-language-server` can be reused in ansible-navigator project to test WSL2 environments in CI.

File: docs/changelog-fragments.d/866.bugfix.md (hunk 1 above)
Comment (webknjaz) [reply docs/changelog-fragments.d/866.bugfix.md]: Looks good, resolving.

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py [hunk 10]
```diff
@@ -281,6 +281,22 @@ def execution_environment(self, entry: SettingsEntry, config) -> PostProcessorRe
                         )
                     if hint:
                         exit_messages.append(ExitMessage(message=hint, prefix=ExitPrefix.HINT))
+
+            # check for /dev/mqueue/ when using podman because runner passes ipc=host
+            # https://github.com/ansible/ansible-navigator/issues/610
+            mqueue_path = "/dev/mqueue/"
+            if config.container_engine == "podman":
+                if not Path(mqueue_path).exists() or not Path(mqueue_path).is_dir():
```
Comment (webknjaz) [inline src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:289]: Plese combine these clauses to dedent the whole conditional block:
```suggestion
            podman_engine_with_no_mqueue_path = (
                config.container_engine == "podman" and
                not Path(mqueue_path).exists() or not Path(mqueue_path).is_dir()
            (
            if podman_engine_with_no_mqueue_path:
```
This way, there's a named conditional with a clearer purpose/semantics + less nesting.

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py [hunk 11]
```diff
@@ -281,6 +281,22 @@ def execution_environment(self, entry: SettingsEntry, config) -> PostProcessorRe
...
+                    exit_msg = (
+                        "Execution environment support while using podman requires a"
+                        f" '{mqueue_path}' directory."
+                    )
+                    exit_messages.append(ExitMessage(message=exit_msg))
+                    hint = (
+                        "Try creating it with 'mkdir {mqueue_path}' or reference the documentation"
```
Comment (webknjaz) [inline src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:296]: Looks like this is missing an f-string prefix.
```suggestion
                        f"Try creating it with 'mkdir {mqueue_path}' or reference the documentation"
```
Is this not covered with tests?

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    :param is_dir: The return value to set for ``pathlib.Path.is_dir``
+    :param ee_support: The value to set for ``--ee``
+    :param engine: The value to set for ``--ce``
+    :param generate_config: The configuration generator fixture
+    """
+    mqueue_msg = (
+        "Execution environment support while using podman requires a '/dev/mqueue/' directory."
+    )
+    with patch("pathlib.Path.exists", return_value=exists):
+        with patch("pathlib.Path.is_dir", return_value=is_dir):
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:28]: Please use the built-in `monkeypatch` fixture — it's a canonical way of doing the same in the pytest ecosystem. FWIW importing `unittest` should be avoided to prevent accidentally limiting pytest's capabilities. Why is it not forbidden via pylint yet?

Comment (webknjaz) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py:28]: Besides, it'd allow for dropping the nesting.

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+from unittest.mock import patch
+
+import pytest
+
+
+@pytest.mark.parametrize("exists,", [True, False], ids=["exists-True", "exists-False"])
+@pytest.mark.parametrize("is_dir,", [True, False], ids=["is_dir-True", "is_dir-False"])
+@pytest.mark.parametrize("ee_support,", [True, False], ids=["ee_support-True", "ee_support-False"])
+@pytest.mark.parametrize("engine,", ["podman", "docker"], ids=["engine-podman", "engine-docker"])
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:11]:
```suggestion
@pytest.mark.parametrize("is_dir", [True, False], ids=["is_dir-True", "is_dir-False"])
@pytest.mark.parametrize("ee_support", [True, False], ids=["ee_support-True", "ee_support-False"])
@pytest.mark.parametrize("engine", ["podman", "docker"], ids=["engine-podman", "engine-docker"])
```

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    :param engine: The value to set for ``--ce``
+    :param generate_config: The configuration generator fixture
+    """
+    mqueue_msg = (
+        "Execution environment support while using podman requires a '/dev/mqueue/' directory."
+    )
+    with patch("pathlib.Path.exists", return_value=exists):
+        with patch("pathlib.Path.is_dir", return_value=is_dir):
+            response = generate_config(params=["--ce", engine, "--ee", str(ee_support)])
+            if ee_support is False or engine == "docker" or (exists is True and is_dir is True):
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:30]: This is quite complex. It's best not to have too much logic in tests as it makes them fragile and hard to follow.
Besides, this complex condition probably has some meaning so I'd put it into a var.

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+                assert mqueue_msg not in [exit_msg.message for exit_msg in response.exit_messages]
+            else:
+                assert mqueue_msg in [exit_msg.message for exit_msg in response.exit_messages]
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:33]: This is basically copy-paste. It could be optimized so that it's not as complex.

File: tests/unit/configuration_subsystem/test_posix_message_queue.py [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+    :param exists: The return value to set for ``pathlib.Path.exists``
+    :param is_dir: The return value to set for ``pathlib.Path.is_dir``
+    :param ee_support: The value to set for ``--ee``
+    :param engine: The value to set for ``--ce``
+    :param generate_config: The configuration generator fixture
+    """
+    mqueue_msg = (
+        "Execution environment support while using podman requires a '/dev/mqueue/' directory."
+    )
```
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:26]:
```suggestion
    )
    mqueue_msg_should_exist = ee_support is False or engine == "docker" or (exists is True and is_dir is True)
```

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 15 above)
Comment (webknjaz) [inline tests/unit/configuration_subsystem/test_posix_message_queue.py:33]: This could easily become
```suggestion
            mqueue_msg_exists = mqueue_msg in [exit_msg.message for exit_msg in response.exit_messages]
            assert mqueue_msg_exists is mqueue_msg_should_exist
```

Comment (webknjaz) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py:33]: Or even better:
```suggestion
            mqueue_msg_exists = any(exit_msg.message == mqueue_msg for exit_msg in response.exit_messages)
            assert mqueue_msg_exists is mqueue_msg_should_exist
```

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 9 above)
Comment (webknjaz) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: I'd stick with this function and not fall into the trap of stdlib unittest.

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 12 above)
Comment (webknjaz) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py:28]: If that patch is really needed (I don't think it is), there's a package called `pytest-mocker` that exposes all of this as a pytest fixture.

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py (hunk 10 above)
Comment (relrod) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:289]: I actually like the original here better because it's easier than having to guess which way Python will associate the set of conditionals. If you do take @webknjaz's suggestion, please wrap the second set of conditionals in parens so it's easy to see at a glance what the actual intent is for this to be parsed.

Comment (webknjaz) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:289]: Also, could just name that second conditional too.

Comment (webknjaz) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:289]: Looking closer, I realized that there's `Path(mqueue_path).exists()` before `is_dir()` — why is it even necessary? Wouldn't `is_dir()` only be true if it exists?

Comment (cidrblock) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py:289]: yep! kinda funny :)

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 16 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py:32]: did something similar

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py (hunk 10 above)
Comment (cidrblock) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py]: @relrod  let me know if I captured your feedback. A lot changed, but I think I got it...

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 12 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: Didn't need it :)

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 14 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: Simpler now I think, LMK

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 15 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: Agree. Done I think

Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: Did something similar, lmk. Good stuff

File: tests/unit/configuration_subsystem/test_posix_message_queue.py (hunk 9 above)
Comment (cidrblock) [reply tests/unit/configuration_subsystem/test_posix_message_queue.py]: This isn't really relevant after the last push. Let me know

**Review by webknjaz — APPROVED** [review]

File: src/ansible_navigator/configuration_subsystem/navigator_post_processor.py (hunk 10 above)
Comment (ssbarnea) [reply src/ansible_navigator/configuration_subsystem/navigator_post_processor.py]: Be warned that is_dir will returns false for symlinked folders. They act as folders but is_dir reports false.
